keps/sig-storage/20200609-csi-service-account-token.md

```diff
@@ -0,0 +1,266 @@
+4.  The token will be specified in `VolumeContext` to `NodePublishVolume` call.
+5.  Every 0.1 second, the reconciler component of volume manager will remount
+    the volume in case the vault secrets expire and re-login is required.
+
+### Notes/Constraints/Caveats
+
+The `RequiresRemount` is useful when the mounted volumes can expire and the
+availability and validity of volumes are continuously required. Those volumes
+are most likely credentials which rotates for the best security practice. There
+are two options when the remount failed:
+
+1.  Keep the container/pod running and use the old credentials.
+    - The next `NodePublishVolume` may succeed if it was unlucky transient
+      failure.
+    - Given there are multiple of 0.1 second usage of stale credentials, it is
+      critical for the credential provisioners to guarantee that the validity
+      is revoked after expiry. In general, it is much harder to eliminate the
+      sinks than source.
+    - The container/pod will also have better observability in usage of the
+      stale credentials.
+2.  Kill the container/pod and hopefully the new container/pod has the refreshed
+    credentials.
+    - This will reduce the stale volume exposure by one sink.
+    - More likely to overcome fatal errors.
+    - Constart-up cost is high
+
+Option 1 is adopted. See discussion
+[here](https://github.com/kubernetes/enhancements/pull/1855#discussion_r443040359).
+
+### Test Plan
+
+- Unit tests around all the added logic in kubelet.
+- E2E tests around remount and token passing.
+
+### Graduation Criteria
+
+#### Alpha->Beta
+
+- Implemented the feature.
+- Wrote all the unit and E2E tests.
+
+#### Beta->GA
+
+- Deployed the feature in production and went through at least minor k8s
+  version.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- **How can this feature be enabled / disabled in a live cluster?**
+
+  - Feature gate name: CSIDriverServiceAccountToken
+  - Components depending on the feature gate: kubelet
+  - Will enabling / disabling the feature require downtime of the control
+    plane? no.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? yes.
+
+- **Does enabling the feature change any default behavior?** no.
+
+- **Can the feature be disabled once it has been enabled (i.e. can we roll
+  back the enablement)?** yes, as long as the new fields in `CSIDriverSpec` is
+  not used.
+
+- **What happens if we reenable the feature if it was previously rolled
+  back?** nothing, as long as the new fields in `CSIDriverSpec` is not used.
+
+- **Are there any tests for feature enablement/disablement?** yes, unit tests
+  will cover this.
+
+### Scalability
+
+- **Will enabling / using this feature result in any new API calls?**
+
+  - API call type: `TokenRequest`
+  - estimated throughput: 1(`RequiresRemount=false`) or
+    10/s(`RequiresRemount=true`) for each CSI driver using this feature.
```

Comment:
Is this a lot? Can it only be called if the token has expired?

Reply:
my bad. it should not be as high as 10/s as it will be cached until the expiration seconds. so it is 1 for ExpirationSeconds.